homeassistant/components/whirlpool/sensor.py

```diff
@@ -0,0 +1,287 @@
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.typing import StateType
+from homeassistant.util.dt import utcnow
+
+from . import WhirlpoolData
+from .const import DOMAIN
+
+TANK_FILL = {
+    "0": "Unknown",
+    "1": "Empty",
+    "2": "25%",
+    "3": "50%",
+    "4": "100%",
+    "5": "Active",
+}
+
+MACHINE_STATE = {
+    MachineState.Standby: "Standby",
+    MachineState.Setting: "Setting",
+    MachineState.DelayCountdownMode: "Delay Countdown",
+    MachineState.DelayPause: "Delay Paused",
+    MachineState.SmartDelay: "Smart Delay",
+    MachineState.SmartGridPause: "Smart Grid Pause",
+    MachineState.Pause: "Pause",
+    MachineState.RunningMainCycle: "Running Maincycle",
+    MachineState.RunningPostCycle: "Running Postcycle",
+    MachineState.Exceptions: "Exception",
+    MachineState.Complete: "Complete",
+    MachineState.PowerFailure: "Power Failure",
+    MachineState.ServiceDiagnostic: "Service Diagnostic Mode",
+    MachineState.FactoryDiagnostic: "Factory Diagnostic Mode",
+    MachineState.LifeTest: "Life Test",
+    MachineState.CustomerFocusMode: "Customer Focus Mode",
+    MachineState.DemoMode: "Demo Mode",
+    MachineState.HardStopOrError: "Hard Stop or Error",
+    MachineState.SystemInit: "System Initialize",
+}
+
+CYCLE_FUNC = [
+    (WasherDryer.get_cycle_status_filling, "Cycle Filling"),
+    (WasherDryer.get_cycle_status_rinsing, "Cycle Rinsing"),
+    (WasherDryer.get_cycle_status_sensing, "Cycle Sensing"),
+    (WasherDryer.get_cycle_status_soaking, "Cycle Soaking"),
+    (WasherDryer.get_cycle_status_spinning, "Cycle Spinning"),
+    (WasherDryer.get_cycle_status_washing, "Cycle Washing"),
+]
+
+
+ICON_D = "mdi:tumble-dryer"
+ICON_W = "mdi:washing-machine"
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def washer_state(washer: WasherDryer) -> str | None:
+    """Determine correct states for a washer."""
+
+    if washer.get_attribute("Cavity_OpStatusDoorOpen") == "1":
+        return "Door open"
+
+    machine_state = washer.get_machine_state()
+
+    if machine_state == MachineState.RunningMainCycle:
+        for func, cycle_name in CYCLE_FUNC:
+            if func(washer):
+                return cycle_name
+
+    return MACHINE_STATE.get(machine_state, STATE_UNKNOWN)
+
+
+@dataclass
+class WhirlpoolSensorEntityDescriptionMixin:
+    """Mixin for required keys."""
+
+    value_fn: Callable
+
+
+@dataclass
+class WhirlpoolSensorEntityDescription(
+    SensorEntityDescription, WhirlpoolSensorEntityDescriptionMixin
+):
+    """Describes Whirlpool Washer sensor entity."""
+
+
+SENSORS: tuple[WhirlpoolSensorEntityDescription, ...] = (
+    WhirlpoolSensorEntityDescription(
+        key="state",
+        name="State",
+        icon=ICON_W,
+        has_entity_name=True,
+        value_fn=washer_state,
+    ),
+    WhirlpoolSensorEntityDescription(
+        key="DispenseLevel",
+        name="Detergent Level",
+        icon=ICON_W,
+        has_entity_name=True,
```

Comment:
Common attributes that are the same for all entities are preferably set on the entity itself.